packages/langium/src/syntax-tree.ts

```diff
@@ -35,6 +35,11 @@ export interface GenericAstNode extends AstNode {
     [key: string]: unknown
 }
 
+export type OptionalAstNode<T extends AstNode> = {
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    [P in keyof T]: P extends keyof AstNode ? T[P] : T[P] extends any[] | boolean ? T[P] : T[P] | undefined;
```

Comment:
```suggestion
    [P in keyof T]: P extends keyof AstNode ? T[P] : T[P] extends unknown[] | boolean ? T[P] : T[P] | undefined;
```
... should do the trick, too, without any need for the linter exclusion.